cmd/notation/verify.go

```diff
@@ -0,0 +1,195 @@
+			TakesFile: true,
+		},
+		&cli.BoolFlag{
+			Name:  "pull",
+			Usage: "pull remote signatures before verification",
+			Value: true,
+		},
+		localFlag,
+		usernameFlag,
+		passwordFlag,
+		plainHTTPFlag,
+		mediaTypeFlag,
+	},
+	Action: runVerify,
+}
+
+func runVerify(ctx *cli.Context) error {
+	// initialize
+	scheme, err := getSchemeForVerification(ctx)
+	if err != nil {
+		return err
+	}
+	manifest, err := getManifestFromContext(ctx)
+	if err != nil {
+		return err
+	}
+
+	sigPaths := ctx.StringSlice(signatureFlag.Name)
+	if len(sigPaths) == 0 {
+		if !ctx.Bool(localFlag.Name) && ctx.Bool("pull") {
+			if err := pullSignatures(ctx, digest.Digest(manifest.Digest)); err != nil {
+				return err
+			}
+		}
+		manifestDigest := digest.Digest(manifest.Digest)
+		sigDigests, err := config.SignatureDigests(manifestDigest)
+		if err != nil {
+			return err
+		}
+		for _, sigDigest := range sigDigests {
+			sigPaths = append(sigPaths, config.SignaturePath(manifestDigest, sigDigest))
+		}
+	}
+
+	// core process
+	if err := verifySignatures(scheme, manifest, sigPaths); err != nil {
+		return err
+	}
+
+	// write out
+	fmt.Println(manifest.Digest)
+	return nil
+}
+
+func verifySignatures(scheme *signature.Scheme, manifest signature.Manifest, sigPaths []string) error {
+	if len(sigPaths) == 0 {
+		return errors.New("verification failure: no signatures found")
+	}
+
+	var lastErr error
+	for _, path := range sigPaths {
+		sig, err := os.ReadFile(path)
+		if err != nil {
+			return err
+		}
+		claims, err := scheme.Verify(string(sig))
+		if err != nil {
+			lastErr = fmt.Errorf("verification failure: %v", err)
+			continue
+		}
+		if manifest.Descriptor != claims.Manifest.Descriptor {
+			lastErr = fmt.Errorf("verification failure: %s", manifest.Digest)
+			continue
+		}
+		return nil
+	}
+	return lastErr
+}
+
+func getSchemeForVerification(ctx *cli.Context) (*signature.Scheme, error) {
+	scheme := signature.NewScheme()
+
+	// add x509 verifier
+	verifier, err := getX509Verifier(ctx)
+	if err != nil {
+		return nil, err
+	}
+	scheme.RegisterVerifier(verifier)
+
+	return scheme, nil
+}
+
+func getX509Verifier(ctx *cli.Context) (signature.Verifier, error) {
+	// resolve paths
+	certPaths := ctx.StringSlice("cert-file")
+	certPaths, err := appendCertPathFromName(certPaths, ctx.StringSlice("cert"))
+	if err != nil {
+		return nil, err
+	}
+	caCertPath := ctx.StringSlice("ca-cert-file")
+	caCertPath, err = appendCertPathFromName(caCertPath, ctx.StringSlice("ca-cert"))
+	if err != nil {
+		return nil, err
+	}
+	if len(certPaths) == 0 && len(caCertPath) == 0 {
+		cfg, err := config.LoadOrDefaultOnce()
+		if err != nil {
+			return nil, err
+		}
+		if len(cfg.VerificationCertificates.Certificates) == 0 {
+			return nil, errors.New("trust certificate not specified")
+		}
+		for _, ref := range cfg.VerificationCertificates.Certificates {
+			certPaths = append(certPaths, ref.Path)
+		}
+	}
+
+	// read cert files
```

Comment:
I think it might be better to move the cert related code to notation-go-lib. The CLI is the shell that only needs to pass the file io handler to the library. It doesn't need to use "crypto/x509" directly.

Reply:
Good point. Will include that in the next major iteration.

Reply:
@northtyphoon Could you create an issue to track?